# 1. Data Analysis

Librerias relevantes para el análisis

In [116]:
library(rio)
library(ggplot2)
library(caTools)
library(DescTools)
library(dplyr)
library(Hmisc)
library(psych)

In [117]:
url="https://raw.githubusercontent.com/d2cml-ai/CausalAI-Course/main/data/wage2015_subsample_inference.csv"
wage<-read.csv(url)

2. Reporte de NAs


In [118]:
colSums(is.na(wage))

rownames     wage    lwage      sex      shs      hsg      scl      clg 
       0        0        0        0        0        0        0        0 
      ad       mw       so       we       ne     exp1     exp2     exp3 
       0        0        0        0        0        0        0        0 
    exp4      occ     occ2      ind     ind2 
       0        0        0        0        0

No existen datos perdidos en la base de datos

### 3.  Reporte de estadísticas descriptivas

In [119]:
describe(wage)

vars n    mean         sd           median       trimmed     
rownames  1   5150 1.563635e+04 9.700744e+03 15260.000000 1.546166e+04
wage      2   5150 2.341041e+01 2.100302e+01    19.230769 2.053559e+01
lwage     3   5150 2.970787e+00 5.703848e-01     2.956512 2.956854e+00
sex       4   5150 4.444660e-01 4.969547e-01     0.000000 4.305825e-01
shs       5   5150 2.330097e-02 1.508723e-01     0.000000 0.000000e+00
hsg       6   5150 2.438835e-01 4.294650e-01     0.000000 1.798544e-01
scl       7   5150 2.780583e-01 4.480858e-01     0.000000 2.225728e-01
clg       8   5150 3.176699e-01 4.656155e-01     0.000000 2.720874e-01
ad        9   5150 1.370874e-01 3.439730e-01     0.000000 4.635922e-02
mw       10   5150 2.596117e-01 4.384641e-01     0.000000 1.995146e-01
so       11   5150 2.965049e-01 4.567606e-01     0.000000 2.456311e-01
we       12   5150 2.161165e-01 4.116346e-01     0.000000 1.451456e-01
ne       13   5150 2.277670e-01 4.194322e-01     0.000000 1.597087e-01
exp1     14   5150 1.376058e+01 1.060947e+01    10.000000 1.261080e+01
exp2     15   5150 3.018925e+00 4.000904e+00     1.000000 2.195211e+00
exp3     16   5150 8.235867e+00 1.448896e+01     1.000000 4.632417e+00
exp4     17   5150 2.511804e+01 5.353022e+01     1.000000 1.087759e+01
occ      18   5150 5.310737e+03 1.187436e+04  4040.000000 3.818353e+03
occ2     19   5150 1.167087e+01 6.966684e+00    13.000000 1.178277e+01
ind      20   5150 6.629155e+03 5.333444e+03  7370.000000 6.724626e+03
ind2     21   5150 1.331689e+01 5.701019e+00    14.000000 1.345170e+01
         mad          min        max          range        skew      
rownames 12622.115100  10.000000 3.264300e+04 32633.000000  0.1445372
wage        10.264154   3.021978 5.288457e+02   525.823695 10.7615297
lwage        0.539897   1.105912 6.270697e+00     5.164785  0.3795262
sex          0.000000   0.000000 1.000000e+00     1.000000  0.2234538
shs          0.000000   0.000000 1.000000e+00     1.000000  6.3180109
hsg          0.000000   0.000000 1.000000e+00     1.000000  1.1924920
scl          0.000000   0.000000 1.000000e+00     1.000000  0.9904294
clg          0.000000   0.000000 1.000000e+00     1.000000  0.7830267
ad           0.000000   0.000000 1.000000e+00     1.000000  2.1097132
mw           0.000000   0.000000 1.000000e+00     1.000000  1.0962889
so           0.000000   0.000000 1.000000e+00     1.000000  0.8908633
we           0.000000   0.000000 1.000000e+00     1.000000  1.3790306
ne           0.000000   0.000000 1.000000e+00     1.000000  1.2978506
exp1         8.895600   0.000000 4.700000e+01    47.000000  0.8243519
exp2         1.349166   0.000000 2.209000e+01    22.090000  1.6435529
exp3         1.470739   0.000000 1.038230e+02   103.823000  2.3217862
exp4         1.482452   0.000000 4.879681e+02   487.968100  2.9923560
occ       2816.940000  10.000000 1.000000e+05 99990.000000  7.4676159
occ2         7.413000   1.000000 2.200000e+01    21.000000 -0.2323432
ind       1912.554000 370.000000 1.000000e+05 99630.000000 13.4106511
ind2         7.413000   2.000000 2.200000e+01    20.000000 -0.1919076
         kurtosis     se          
rownames  -1.21638579 1.351766e+02
wage     210.91873662 2.926699e-01
lwage      1.14948916 7.948118e-03
sex       -1.95044702 6.924894e-03
shs       37.92462584 2.102354e-03
hsg       -0.57807485 5.984448e-03
scl       -1.01924743 6.243923e-03
clg       -1.38713847 6.488194e-03
ad         2.45136580 4.793146e-03
mw        -0.79830548 6.109848e-03
so        -1.20659669 6.364804e-03
we        -0.09829356 5.735989e-03
ne        -0.31564500 5.844645e-03
exp1      -0.37320088 1.478393e-01
exp2       1.94297633 5.575124e-02
exp3       5.35155309 2.018988e-01
exp4      10.17782024 7.459255e-01
occ       56.67967275 1.654651e+02
occ2      -1.32512585 9.707838e-02
ind      234.27639832 7.431973e+01
ind2      -1.21131213 7.944176e-02

**Interpretación** : se puede extraer información relevante si nos enfocamos en la variable del salario del individuo. Se puede observar que el salario promedio de la muestra es de 23.41 dólares.No obstante, la desviación estándar es muy alta(21), lo cual quiere decir que existe una alta dispersión del salario con respecto al promedio. Por ejemplo, existe un individuo cuyo salario es de 528.85 dólares y otro cuyo salario es de 3.02 dólares. En aras de obtener información más estable sobre el salario se recurre a la mediana, la cual nos dice que el 50% de la muestra posee un salario entre los 3.02 y 19.23 dólares. Por último, la muestra presenta una asimetría positiva, lo cual implica que existe un mayor número de trabajadores con un salario menor al promedio.

4. Obtener el número de mujeres con un grado académico alto que poseen un salario del 25% más rico de la muestra

4.1 Identificación de cuantiles

In [120]:
summary(wage$wage)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  3.022  13.461  19.231  23.410  27.778 528.846 

4.2 Creamos nueva variable

In [121]:
wage$wage_ordinal=cut(wage$wage, breaks = c(0,13.461,19.231,27.778,528.846),
                      include.lowest = T, ordered_result = T,
                      labels=c("Muy bajo", "Bajo", "Alto","Muy alto"))
table(wage$wage_ordinal)


Muy bajo     Bajo     Alto Muy alto 
    1269     1462     1132     1287 

4.3 Creamos dos subsets que solo contengan a uno de los sexos

In [122]:
wage_only_women=wage%>%filter(sex == 1)
wage_only_men= wage%>%filter(sex == 0)
wage_only_women$case<- c(1)
wage_only_men$case<- c(1)

4.4 Agrupamos

In [123]:
wage_only_women%>%group_by(clg,wage_ordinal)%>%summarise(Casos=sum(case))

`summarise()` has grouped output by 'clg'. You can override using the `.groups`
argument.


clg wage_ordinal Casos
1 0   Muy bajo     484  
2 0   Bajo         434  
3 0   Alto         281  
4 0   Muy alto     295  
5 1   Muy bajo     107  
6 1   Bajo         227  
7 1   Alto         224  
8 1   Muy alto     237


Existen 237 mujeres con un college dregee que poseen un salario del 25% más rico de la muestra.

In [124]:
wage_only_women%>%group_by(ad,wage_ordinal)%>%summarise(Casos=sum(case))

`summarise()` has grouped output by 'ad'. You can override using the `.groups`
argument.


ad wage_ordinal Casos
1 0  Muy bajo     558  
2 0  Bajo         581  
3 0  Alto         392  
4 0  Muy alto     357  
5 1  Muy bajo      33  
6 1  Bajo          80  
7 1  Alto         113  
8 1  Muy alto     175

Existen 175 mujeres con un advanced degree que poseen un salario del 25% más rico de la muestra. En ambos dataframes, las mujeres que no poseen alguno de estos grados académicos tienden a ser más propensas a tener un salario del 25% menos rico de la muestra, en comparación a las mujeres que sí poseen un mayor nivel académico. Ello nos lleva a inferir que probablemente el nivel educativo puede influir en la obtención de mejores salarios.

5. Obtener el número de hombres con un grado académico medio-bajo que poseen un salario perteneciente al 25% más rico de la muestra

In [125]:
wage_only_men%>%group_by(hsg, wage_ordinal)%>%summarise(Casos=sum(case))

`summarise()` has grouped output by 'hsg'. You can override using the `.groups`
argument.


hsg wage_ordinal Casos
1 0   Muy bajo     369  
2 0   Bajo         525  
3 0   Alto         477  
4 0   Muy alto     648  
5 1   Muy bajo     309  
6 1   Bajo         276  
7 1   Alto         150  
8 1   Muy alto     107

Existen 107 trabajadores hombres con secundaria completa que poseen un salario perteneciente al 25% más rico de la muestra.

In [126]:
wage_only_men%>%group_by(shs, wage_ordinal)%>%summarise(Casos=sum(case))

`summarise()` has grouped output by 'shs'. You can override using the `.groups`
argument.


shs wage_ordinal Casos
1 0   Muy bajo     634  
2 0   Bajo         775  
3 0   Alto         614  
4 0   Muy alto     747  
5 1   Muy bajo      44  
6 1   Bajo          26  
7 1   Alto          13  
8 1   Muy alto       8

Existen solo 8 trabajadores hombres con secundaria incompleta que tienen un salario perteneciente al 25% más rico de la muestra. Para ambos casos, los hombres que no poseen alguno de estos grados educativos son más propensos a poseer un salario pertenenciente al 25% más pobre. Si comparamos estos resultados con el anterior grupo, podemos observar que hay más mujeres con un salario altísimo en comparación con los hombres, pero ello puede deberse al mayor nivel educativo que poseen.

6. Creación de dos bases de datos

In [127]:
Y=wage$lwage

8. Hacer tres matrices para nuestros predictores 

8.1

In [128]:
X_1 <- model.matrix(~ sex + hsg + scl + clg + ad + so + we + ne + exp1 + C(occ2) + C(ind2), data = wage)

8.2

In [129]:
X_2 <- model.matrix(~ sex + (exp1 + exp2 + exp3 + exp4) * (hsg + scl + clg + ad + so + we + ne + C(occ2) + C(ind2)), data = wage)

8.3

In [130]:
X_3<- model.matrix(~ sex + (exp1 + exp2 + exp3 + exp4 + hsg + scl + clg + ad + so + we + ne + C(occ2) + C(ind2)) ^ 2, data = wage)

# 3. Linear Regressions 

9. 

In [131]:
train_sample <- runif(length(Y)) < 0.8
test_sample <- !train_sample

10.

In [132]:
modelo_1 <- lm(lwage ~ sex + hsg + scl + clg + ad + so + we + ne + exp1 + C(occ2) + C(ind2), data = wage, subset = train_sample)

modelo_2 <- lm(lwage ~ sex + (exp1 + exp2 + exp3 + exp4) * (hsg + scl + clg + ad + so + we + ne + C(occ2) + C(ind2)), data = wage, subset = train_sample)

modelo_3 <- lm(lwage ~ sex + (exp1 + exp2 + exp3 + exp4 + hsg + scl + clg + ad + so + we + ne + C(occ2) + C(ind2)) ^ 2, data = wage, subset = train_sample)

11.

In [133]:
mse_train_1 <- mean(modelo_1$residuals ^ 2)
r2_train_1 <- 1 - mse_train_1 / var(Y[train_sample])
adjr2_train_1 <- 1 - length(Y) / (length(Y) - ncol(X_1)) * mse_train_1 / var(Y)
mse_test_1 <- mean((predict(modelo_1, subset(wage, subset = test_sample)) - Y[test_sample]) ^ 2)
r2_test_1 <- 1 - mse_test_1 / var(Y[test_sample])

print(paste("Training MSE for basic model:", mse_train_1))
print(paste("Training R2 for basic model:", r2_train_1))
print(paste("Adjusted training R2 for basic model:", adjr2_train_1))
print(paste("Testing MSE for basic model:", mse_test_1))
print(paste("Testing R2 for basic model:", r2_test_1))

Warning message:
"contrasts dropped from factor C(occ2)"
Warning message:
"contrasts dropped from factor C(ind2)"


[1] "Training MSE for basic model: 0.227911222240533"
[1] "Training R2 for basic model: 0.305010100067642"


[1] "Adjusted training R2 for basic model: 0.292458259016196"
[1] "Testing MSE for basic model: 0.214409705587151"
[1] "Testing R2 for basic model: 0.321069766559237"


In [134]:
mse_train_2 <- mean(modelo_2$residuals ^ 2)
r2_train_2 <- 1 - mse_train_2 / var(Y)
adjr2_train_2 <- 1 - length(Y) / (length(Y) - ncol(X_2)) * mse_train_2 / var(Y)
mse_test_2 <- mean((predict(modelo_2, subset(wage, subset = test_sample)) - Y[test_sample]) ^ 2)
r2_test_2 <- 1 - mse_test_2 / var(Y[test_sample])

print(paste("Training MSE for flexible model:", mse_train_2))
print(paste("Training R2 for flexible model:", r2_train_2))
print(paste("Adjusted training R2 for flexible model:", adjr2_train_2))
print(paste("Testing MSE for flexible model:", mse_test_2))
print(paste("Testing R2 for flexible model:", r2_test_2))

Warning message:
"contrasts dropped from factor C(occ2)"
Warning message:
"contrasts dropped from factor C(ind2)"


[1] "Training MSE for flexible model: 0.211745710212049"
[1] "Training R2 for flexible model: 0.349153221883582"
[1] "Adjusted training R2 for flexible model: 0.316504708951967"
[1] "Testing MSE for flexible model: 0.228068223993062"
[1] "Testing R2 for flexible model: 0.277819946946892"


In [135]:
mse_train_3 <- mean(modelo_3$residuals ^ 2)
r2_train_3 <- 1 - mse_train_3 / var(Y)
adjr2_train_3 <- 1 - length(Y) / (length(Y) - ncol(X_3)) * mse_train_3 / var(Y)
mse_test_3 <- mean((predict(modelo_3, subset(wage, subset = test_sample)) - Y[test_sample]) ^ 2)
r2_test_3 <- 1 - mse_test_3 / var(Y[test_sample])

print(paste("Training MSE for flexible model:", mse_train_3))
print(paste("Training R2 for flexible model:", r2_train_3))
print(paste("Adjusted training R2 for flexible model:", adjr2_train_3))
print(paste("Testing MSE for flexible model:", mse_test_3))
print(paste("Testing R2 for flexible model:", r2_test_3))

Warning message:
"contrasts dropped from factor C(occ2)"


Warning message:
"contrasts dropped from factor C(ind2)"
Warning message in predict.lm(modelo_3, subset(wage, subset = test_sample)):
"prediction from rank-deficient fit; attr(*, "non-estim") has doubtful cases"


[1] "Training MSE for flexible model: 0.173815415819514"
[1] "Training R2 for flexible model: 0.46574028224795"
[1] "Adjusted training R2 for flexible model: 0.340182842584399"
[1] "Testing MSE for flexible model: 0.275103938512049"
[1] "Testing R2 for flexible model: 0.128880939960341"


Interpretación: 
- En términos generales, los tres modelos presentan métricas de predicción más altas en el conjunto de entrenamiento, en comparación cuando se evalúan en el conjunto de prueba. 
- En el conjunto de entrenamiento, se puede observar que el modelo flexible tiene un mayor r cuadrado(0.327) en comparación con el modelo básico y el modelo más flexible. Esta tendencia se mantiene incluso cuando utilizamos el r cuadrado (0.465), el cual incluye una penalidad por la inclusión de variables irrelevantes. 
- Cuando empleamos el modelo en el conjunto de prueba, vemos que el modelo más flexible es el que posee mayor error de predicción que los otros modelos (0.275>0.228>0.214), así como un menor R2(0.129<0.278<0.321), ya que incluye una gran cantidad de variables irrelevantes.
- Vemos que el mejor modelo en términos de predicción y de poder explicativo es el modelo básico, con un MSE de 0.214 y un R2 de 0.321.  